xapian-core/api/queryinternal.cc

```diff
@@ -865,98 +867,156 @@ Query::Internal::unserialise(const char ** p, const char * end,
-			throw SerialisationError("not enough data");
-		    }
-		    int flags = static_cast<unsigned char>(*(*p)++);
-		    op combiner = static_cast<op>(*(*p)++);
-		    unsigned edit_distance;
-		    size_t fixed_prefix_len;
-		    string pattern;
-		    if (!unpack_uint(p, end, &edit_distance) ||
-			!unpack_uint(p, end, &fixed_prefix_len) ||
-			!unpack_string(p, end, pattern)) {
-			throw SerialisationError("not enough data");
+	    switch (ch >> 4) {
+	    case 1: {
+		// 0001oxxx where:
+		// o is the operation: LT or GT
+		// xxx is the slot number
+		Xapian::valueno slot = ch & 7;
+		if (slot == 7) {
+		    if (!unpack_uint(p, end, &slot)) {
+			unpack_throw_serialisation_error(*p);
+		}
+		slot += 7;
+		}
+		string begin;
+		if (!unpack_string(p, end, begin)) {
+		    unpack_throw_serialisation_error(*p);
+		}
+		switch (ch >> 3)
+		{
+		case 2: {
+		    // 00010xxx for OP_VALUE_LT
+		    return new Xapian::Internal::QueryValueGT(slot,
+							      begin);
+		}
+		case 3: {
+		    // 00011xxx for OP_VALUE_GT
+		    string end_;
+		    if (!unpack_string(p, end, end_)) {
+			unpack_throw_serialisation_error(*p);
 		    }
-		    using Xapian::Internal::QueryEditDistance;
-		    return new QueryEditDistance(pattern,
-						 max_expansion,
-						 flags,
-						 combiner,
-						 edit_distance,
-						 fixed_prefix_len);
+		return new Xapian::Internal::QueryValueLT(slot, end_);
+		}
 		}
-		case 0x0b: { // Wildcard
-		    Xapian::termcount max_expansion;
-		    if (!unpack_uint(p, end, &max_expansion) || end - *p < 2) {
+		break;
+	    }
+
+	    case 0: {
+		// Other operators
+		//
+		//   0000tttt where:
+		//     tttt -> encodes which OP_XXX
+		switch (ch & 0x0f) {
+		    case 0x00: // OP_INVALID
+			return new Xapian::Internal::QueryInvalid();
+		    case 0x0a: { // Edit distance
+			Xapian::termcount max_expansion;
+			if (!unpack_uint(p, end, &max_expansion) ||
+						 end - *p < 2) {
 			throw SerialisationError("not enough data");
+			}
+			int flags =
+				static_cast<unsigned char>(*(*p)++);
+			op combiner = static_cast<op>(*(*p)++);
+			unsigned edit_distance;
+			size_t fixed_prefix_len;
+			string pattern;
+			if (!unpack_uint(p, end,
+					 &edit_distance) ||
+				!unpack_uint(p, end,
+					 &fixed_prefix_len) ||
+				!unpack_string(p, end,
+						pattern)) {
+			    throw SerialisationError(
+						"not enough data");
+			}
+			using Xapian::Internal::QueryEditDistance;
+			return new QueryEditDistance(pattern,
+						     max_expansion,
+						     flags,
+						     combiner,
+						     edit_distance,
+						     fixed_prefix_len);
 		    }
-		    int flags = static_cast<unsigned char>(*(*p)++);
-		    op combiner = static_cast<op>(*(*p)++);
-		    string pattern;
-		    if (!unpack_string(p, end, pattern)) {
+		    case 0x0b: { // Wildcard
+			Xapian::termcount max_expansion;
+			if (!unpack_uint(p, end,
+					 &max_expansion) ||
+					 end - *p < 2) {
+				throw SerialisationError(
```

Comment:
I think this is overindented (it should be four spaces to the right of the containing `if()`).